### PR TITLE
Add option for fanning out FetchDocuments to many workers

### PR DESCRIPTION
**Pull request by zph**
Thank you for creating this project :). It's quite helpful in me building https://github.com/zph/moresql.

This PR adds ability to set option of using many workers to process the FetchDocuments function.

I was digging into my own code to find a performance bottleneck and found it's in GTM. Seems like `FetchDocuments` cannot keep up well with multiple documents being updated at once.

Rough benchmark (in https://github.com/zph/moresql project) of 140 Mongo documents being updated near simultaneously while tailing w/ GTM:
* Existing code results in ~13 second lag by the time last record is processed
* With patch, when fanned out to 50 workers, maximum lag stays at ~1.3 seconds.

GTM settings by default retain the existing behavior of 1 worker so that library keeps backwards compatibility.

Does this patch seem like something you're interested in building into rwynn/gtm?

Regards,
Zander

### CODE REVIEW

**Comment by rwynn** [comment]
Hi @zph ,

Thanks for taking gtm for a spin.  Your moresql project looks pretty cool.  I’ll have to give that a try at some point.  
I thought a little about this PR last night.  At first I thought there should be no problems and that your approach would big a big performance win.  However, then after thinking a bit more I began to have some reservations.  

I think with more than 1 Worker that semantics of tailing the oplog may be lost in some situations.  Consider an insert of a document followed quickly by a delete of the document.  With multiple workers it is possible that the ordering of the events is switched coming out of the output channel in gtm.  This is due to how gtm processes inserts and deletes differently (and the fact that we have more than 1 worker).  For deletes the op is quickly passed to the output channel, but for Inserts/Updates gtm tries to retrieve the document before sending it out on the output channel.  So, it is likely that the delete would be sent in front of the insert.  This would break the semantics that are currently in place where events are sent in oplog order on the output channel. 

This becomes a problem for projects that use gtm to replicate changes to other systems.  For instance, in my project [monstache](https://github.com/rwynn/monstache) the above example of an insert followed quickly by a delete might result in a document getting indexed into Elastic Search where it should have been deleted.   It’s possible that this edge case may affect your project as well.  

I can think of 2 scenarios where out of order events might cause problems: deletes and table/collection drops.  Monstache, will for example, handle a drop by deleting an index in Elastic search.

I still think that we can improve throughput for your use case (lots of documents updated at once) while still maintaining in order output on the channel (keeping only 1 worker).   

I was planning to take a shot today at an approach where the ops are buffered such that instead of doing a FindId for each document, instead the buffer would get filled up to a certain point and then the buffer would be flushed (meaning that a single Find would be issued to mongo with a query with multiple ids in an $in clause).   

So, for example, an update in mongodb that affects 100 documents might result in 2 finds of 50 document ids at once instead of 100 finds with 1 id at once.  I’ll let you know how this goes.  I plan to support a timeout such that if the buffer is not filled in a configurable amount of time (e.g. 5 seconds) then a buffer is forced to flush.  So the buffer flushes it is reaches a number of documents or if a short period of time elapses.

**Comment by zph** [comment]
Dear @rwynn,

tl;dr - Agreed and I'm interested in the batching approach you mention with 
* Flush count - max number of operations to buffer before making batch request
* Timer flush duration - max `time.Duration` between each batch request

I like the idea of bulk fetching and then ensuring they're sorted by MongoTimestamp before placing them on the out channel.

I wasn't considering the speed difference of INSERT vs DELETE but did realize there's a potential loss of ordering semantics even within INSERT/UPDATE depending on network response for each. In moresql for our use case it makes sense where speed is more of a priority than full consistency. For users needing higher consistency, moresql allows for bottlenecking worker counts internally.

Thanks for considering this general improvement useful and working on an implementation. If you're unable to dedicate the time to work on it, please let me know and I'll contribute :).

**Comment by rwynn** [comment]
Hi @zph,
Just created an issue, #4, which a gtm.go file attached to it.  Please give that a shot and let me know if it works out for you.  The public API is the same, but some new gtm.Options have been added.  The default batch size is 50 (MaxDocs) and the default duration is 5s (MaxDuration).   Hopefully, by playing with these you can optimize for your workload.

**Comment by rwynn** [comment]
Closing PR #3 since receiving feedback that the fix for Issue #4  improves performance while maintaining ordering semantics.

**Comment by zph** [comment]
@rwynn Thanks for solving this :).

**Comment by rwynn** [comment]
@zph,
Thanks for identifying and working with me on this one. Looking forward to hearing about the official release of moresql.

**Comment by rwynn** [comment]
@zph,

[gtm.zip](https://github.com/rwynn/gtm/files/771642/gtm.zip)
Hello again :-).  Reworked gtm one last time to hopefully squeeze out a little more performance for you.  Please try the attached version which adds WorkerCount and Ordering options. This is a riff off your original PR which tries to combine both fan out and bulk fetch approaches while still being explicit about the ordering.  This should let you relax the ordering guarantee which then lets you fan out the fetch.  

You can choose to maintain oplog order within a namespace or single document history.  

I'd be interested to see if this helps in your benchmark test.

**Comment by zph** [comment]
@rwynn,

I tried it out today on a similar dataset and compared existing performance to this patch (with Document level consistency).

Results in an HTML page here that shows quantiles of what percentage of responses during that time occured within X ms.

http://data.xargs.io/20170213165151-performance.html

It appears that using Document level consistency during this small test demonstrated that 50% of results were received with ~670ms or less. When done with previous implementation of bulk fetching, the 50% value was 1011ms. Past the 50% cutoff their performance was much more similar.

Seems like a win from my perspective :). And I enjoyed reading the implementation using "github.com/serialx/hashring". I didn't realize this mechanism existed in Golang (https://github.com/zph/gtm/blob/try-out-gtm-with-types-of-ordering/gtm.go#L23-L25) where they're assigned sequential numbers.

If you want to do more collaboration, feel free to put up a branch and point me at that. It's a bit easier to integrate into trial project :).

**Comment by zph** [comment]
Please give me a shout once you integrate it into the project and I'll upgrade moresql to use the newer semantics.

**Comment by rwynn** [comment]
@zph,
Awesome work analyzing this change.  I've committed it.  

I was happy to learn about iota, too.  I'm a big Go fan at this point which it sounds like we have in common.  

I would love to know the details behind how you created those graphs and that analysis.  Using R for that?

Yes, a git feature branch is the way I should have gone.  Will go that route in the future for this sort of thing.  I'm still a little hesitant to use even the most basic "advanced" features of git for fear of shooting myself in the foot.  It's probably easy enough.

**Comment by zph** [comment]
😺 

I'm becoming more and more of a Golang fan... I tend to be a little agnostic but go's solving the problems I'm interested in lately. I'm not thrilled by the language constructs in go (beyond concurrency) but I am thrilled by what I can accomplish. It feels like I can focus on the outcome versus the fanciness of implementation :).

For the R stuff, it's something our Analytics team exposed me to and I can do very crude things like this 😉 . I used RStudio (open source IDE) and built it as an RMarkdown file: https://gist.github.com/zph/300fd1008c6ad29e089eac244f9601c1. Included are the datafiles and the RMarkdown. It allows you to have executable codeblocks that will output results as graphics. (Any inaccuracies are mine and not the Analyst's fault. They didn't review this iteration of results.)

Thinking about the git branch, it's odd when less familiar. But if you can build out `gtm` you'll be fine w/ Git :). `git checkout -b <new-branch-name>` to create it, then `git push` to send it up to Github. When it's merged into Github, `git checkout master && git pull master` to get those changes. The Github GUI is decent for this stuff too if you don't want to remember the cmdline args.

Thank you again :). I enjoyed reading and learning about consistent hashing algos thanks to your commits on gtm.

@ZPH
